src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs

```diff
@@ -1041,102 +1041,184 @@ private bool IsValidFixedVariableInitializer(TypeSymbol declType, SourceLocalSym
+
+                        Error(diagnostics, ErrorCode.ERR_ExprCannotBeFixed, initializerSyntax);
                     }
+                    finally
+                    {
+                        additionalDiagnostics.Free();
+                    }
+
+                    return false;
+            }
+
+            if (elementType.IsManagedType)
+            {
+                Error(diagnostics, ErrorCode.ERR_ManagedAddr, initializerSyntax, elementType);
+                hasErrors = true;
+            }
+
+            initializerOpt = GetFixedLocalCollectionInitializer(initializerOpt, elementType, declType, getPinnableMethod, hasErrors, diagnostics);
+            return true;
+        }
+
+        private MethodSymbol GetDangerousGetPinnableReferenceMethod(BoundExpression initializer, DiagnosticBag additionalDiagnostics)
+        {
+            if (initializer.Type.SpecialType == SpecialType.System_Void)
+            {
+                return null;
+            }
+
+            const string methodName = "DangerousGetPinnableReference";
+
+            DiagnosticBag bindingDiagnostics = DiagnosticBag.GetInstance();
+            try
+            {
+                var boundAccess = BindInstanceMemberAccess(initializer.Syntax, initializer.Syntax, initializer, methodName, rightArity:0, typeArgumentsSyntax: default, typeArguments: default, invoked:true, bindingDiagnostics);
+
+                if (boundAccess.Kind != BoundKind.MethodGroup)
+                {
+                    // the thing is not even a method
+                    return null;
                 }
-                else
+
+                var analyzedArguments = AnalyzedArguments.GetInstance();
+                BoundExpression getPinnableReferenceCall = BindMethodGroupInvocation(initializer.Syntax, initializer.Syntax, methodName, (BoundMethodGroup)boundAccess, analyzedArguments, bindingDiagnostics, queryClause: null, allowUnexpandedForm: false);
+                analyzedArguments.Free();
+
+                if (getPinnableReferenceCall.Kind != BoundKind.Call)
                 {
-                    // convert to generate any additional conversion errors
-                    initializerOpt = GenerateConversionForAssignment(declType, initializerOpt, diagnostics);
-                    return true;
+                    // did not find anythig callable
+                    return null;
                 }
 
-                initializerOpt = GetFixedLocalCollectionInitializer(initializerOpt, elementType, declType, hasErrors, diagnostics);
-            }
+                var call = (BoundCall)getPinnableReferenceCall;
+                if (call.ResultKind == LookupResultKind.Empty)
+                {
+                    // did not find any methods that even remotely fit
+                    return null;
+                }
 
-            return true;
+                var getPinnableReferenceMethod = call.Method;
+                if (getPinnableReferenceMethod is ErrorMethodSymbol ||
```

Comment:
Can we avoid testing a specific type?

Reply:
This is how we test for error method symbol in other places. 

---
In reply to: [167652407](https://github.com/dotnet/roslyn/pull/24469#discussion_r167652407) [](ancestors = 167652407)